bson/src/main/org/bson/codecs/configuration/CompoundCodecRegistry.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.configuration;
+
+import org.bson.codecs.Codec;
+
+import static org.bson.assertions.Assertions.notNull;
+
+final class CompoundCodecRegistry implements CodecRegistry {
+    private final CodecRegistry firstCodecRegistry;
+    private final CodecRegistry secondCodecRegistry;
+
+    CompoundCodecRegistry(final CodecRegistry firstCodecRegistry, final CodecRegistry secondCodecRegistry) {
+        this.firstCodecRegistry = notNull("firstCodecRegistry", firstCodecRegistry);
+        this.secondCodecRegistry = notNull("secondCodecRegistry", secondCodecRegistry);
+    }
+
+    @Override
+    public <T> Codec<T> get(final Class<T> clazz) {
+        try {
+            return firstCodecRegistry.get(clazz);
+        } catch (CodecConfigurationException e) {
```

Comment:
This is related to my comment below about the codecs cache in ProvidersCodecRegistry.  Imagine a situation where the secondCodecRegistry contains most of the codecs (which is probably the case in normal usage).  Now every single call to get will result in a catch of CodecConfigurationException.  Since exceptions are typically quite a bit slower that normal control flow, this could become a performance bottleneck.  I think we either need a separate codecs cache in this class, or else introduce a tryGet method to CodecRegistry.  Of the two, I prefer the first option as it's a simpler design.